File: Olympiad/POI/official/2004/code/mis/graj.pas
{
// Zadanie MIS
// program generujacy rozgrywke i zwracajacy jej wynik, wersja PAS
// Gracz Lolek gra optymalnie, gracz Bolek jest testowany
// Lukasz Kowalik, 14.02.2004
// Poprawki: Rafal Rusin, 2004.03.16
}
program Graj;

uses
  mis; {tu mozna podmieniac testowanego gracza np. na misb1}


{Zawodnik Lolek - Identyczny z wzorcowym Bolkiem}
const
  MAX_MISIE 		=	10000000;
  MAX_MPO 			=	1000000;
  UNDEF					= -1;

var
  _mpo : LongInt;

  PrzegrKol : array [0..MAX_MPO] of LongInt;
  PrzegrPrz : array [-MAX_MPO..MAX_MPO] of LongInt;

{Niech pula bedzie wygrywajaca gdy Bolek moze wygrac niezaleznie od ruchu Lolka
i przegrywajaca w przec. przypadku.
Procedura poczatek wypelnia PrzegrKol i PrzegrPrz tak ze:
PrzegrKol[k]=w   gdy   pula w misiow i k patysiow jest przegrywajaca
PrzegrPrz[p]=k   gdy   pula k+p misiow i k patysiow jest przegrywajaca

W ten sposob:
* z PrzegrKol[k] mozemy odczytac pole przegrywajace w kolumnie k,
* z PrzegrKol[w] mozemy odczytac pole przegrywajace w wierszu w,
* z PrzegrPrz[p] mozemy odczytac pole przegrywajace w przekatnej p}
procedure poczatek_lolka (m, p, mpo : LongInt);
var
  Kolumna, Przekatna : LongInt;
begin
  _mpo := mpo;

  for Kolumna := 0 to mpo do
    PrzegrKol [Kolumna] := UNDEF;

  for Kolumna := -mpo to mpo do
    PrzegrPrz [Kolumna] := UNDEF;

  PrzegrKol [0] := 0;
  PrzegrPrz [0] := 0;
  Przekatna := 1;

  Kolumna := 1;

  while Kolumna + Przekatna <= mpo do
  begin
    if PrzegrKol [Kolumna] = UNDEF then
    begin
      PrzegrKol [Kolumna] := Kolumna + Przekatna;
      PrzegrKol [Kolumna + Przekatna] := Kolumna;
      PrzegrPrz [Przekatna] := Kolumna;
      PrzegrPrz [-Przekatna] := Kolumna + Przekatna;
      Inc (Przekatna);
    end;
    Inc (Kolumna);
  end;

end;


procedure ruch_lolka (m, p : LongInt; var bm, bp : LongInt);
var
  w, k : LongInt;
begin
  w := m mod (_mpo + 1);
  k := p mod (_mpo + 1);

  if (PrzegrKol [k] <> UNDEF) and (w > PrzegrKol [k]) then
  begin
    bm := w - PrzegrKol [k];
    bp := 0;
  end
  else
  if (PrzegrKol [w] <> UNDEF) and (k > {PrzegrWie [w] = } PrzegrKol [w]) then
  begin
    bp := k - PrzegrKol [w];
    bm := 0;
  end
  else
  if (PrzegrPrz [w - k] <> UNDEF) and (k > PrzegrPrz [w - k]) then
  begin
    bm := k - PrzegrPrz [w - k];
    bp := bm;
  end
  else
  begin
    if m > 0 then begin bm := 1; bp := 0; end else begin bp := 1; bm := 0; end;
  end;
end;
{---}

var
  m, p, mpo : LongInt;
  bm, bp, lm, lp : LongInt;

begin
  ReadLn (m, p, mpo);

  poczatek (m, p, mpo);
  poczatek_lolka(m,p,mpo);

  while True do
  begin
    ruch_bolka (m, p, bm, bp);

    m := m - bm;
    p := p - bp;
    //WriteLn ('bm = ', bm, '; bp = ', bp);
    if (m < 0) or (p < 0) or
       ((bm > 0) and (bp > 0) and (bm <> bp)) or
       (bm > mpo) or (bp > mpo) or
       ((bm = 0) and (bp = 0)) then
    begin
      WriteLn ('WRONG');
      Halt;
    end;

    if (m = 0) and (p = 0) then
    begin
      WriteLn ('OK');
      Halt;
    end;

{    WriteLn ('m = ', m, '; p = ', p);
    Write ('> ');
    ReadLn (lm, lp);}

    ruch_lolka (m, p, lm, lp); //gra przeciwko optymalnemu lolkowi
    //WriteLn ('lm = ', lm, '; lp = ', lp);

    m := m - lm;
    p := p - lp;

    if (m = 0) and (p = 0) then
    begin
      WriteLn ('WRONG');
      Halt;
    end;
  end;

end.

File: Olympiad/POI/official/2004/code/mis/graj.c
// Zadanie MIS
// program generujacy rozgrywke i zwracajacy jej wynik, wersja C/C++
// Gracz Lolek gra optymalnie, gracz Bolek jest testowany
// Lukasz Kowalik, 14.02.2004
// Poprawki: Rafal Rusin, 2004.03.16

#include "mis.h"
#include <stdio.h>
#include <stdlib.h>

#define UNDEF -1

static int m, p, mpo;
static int bm, bp, lm, lp;

static int przegr_kol [MAX_MPO+1];
static int przegr_prz [2*MAX_MPO+1];

void inicjuj (int m, int p, int mpo)
{
  int kolumna, przekatna;
  int i;

  for (kolumna = 0; kolumna <= mpo; kolumna++)
    przegr_kol [kolumna] = UNDEF;

  for (i = -mpo; i <= mpo; i++)
    przegr_prz [i + mpo] = UNDEF;

  przegr_kol [0] = 0;
  przegr_prz [0 + mpo] = 0;
  przekatna = 1;

  kolumna = 1;

  while (kolumna + przekatna <= mpo)
  {
    if (przegr_kol [kolumna] == UNDEF)
    {
      przegr_kol [kolumna] = kolumna + przekatna;
      przegr_kol [kolumna + przekatna] = kolumna;
      przegr_prz [przekatna + mpo] = kolumna;
      przegr_prz [-przekatna + mpo] = kolumna + przekatna;
      przekatna++;
    }
    kolumna++;
  }
}

void ruch_lolka (int m, int p, int *lm, int *lp)
{
  int w, k;

  w = m % (mpo + 1);
  k = p % (mpo + 1);

  if ((przegr_kol [k] != UNDEF) && (w > przegr_kol [k]))
  {
    *lm = w - przegr_kol [k];
    *lp = 0;
  }
  else
  if ((przegr_kol [w] != UNDEF) && (k > /*przegr_wie [w] = */ przegr_kol [w]))
  {
    *lp = k - przegr_kol [w];
    *lm = 0;
  }
  else
  if ((przegr_prz [w - k + mpo] != UNDEF) && (k > przegr_prz [w - k + mpo]))
  {
    *lm = k - przegr_prz [w - k + mpo];
    *lp = *lm;
  }
  else
    /* pozycja przegrywajaca, robimy cokolwiek */
    if (m > 0) { *lm = 1; *lp = 0; } else { *lp = 1; *lm = 0; }
}



int main ()
{
  scanf("%d%d%d", &m, &p, &mpo);

  poczatek (m, p, mpo);
  inicjuj (m, p, mpo);

  while (1)
  {
    ruch_bolka (m, p, &bm, &bp);

    m -= bm;
    p -= bp;

    if ((m < 0) || (p < 0) ||
       ((bm > 0) && (bp > 0) && (bm != bp)) ||
       (bm > mpo) || (bp > mpo) ||
       ((bm == 0) && (bp == 0)))
    {
      printf ("WRONG\nNiepoprawny ruch!\n");
      exit(0);
    }

    if ((m == 0) && (p == 0))
    {
      printf ("OK\n");
      exit(0);
    }

    //optymalna strategia Lolka
    ruch_lolka (m, p, &lm, &lp);

    m -= lm;
    p -= lp;

    if ((m == 0) && (p == 0))
    {
      printf ("WRONG\nPrzegrana Bolka!\n");
      exit(0);
    }
  }
}


